### PR TITLE
Feature/k path

### PR DESCRIPTION
**Pull request by mchacki**
### Scope & Purpose

Adds a new AQL functionality called "K_PATHS" that allows to enumerate all paths between a "source" and a "target" vertex of a defined length (min .. max) 

- [ ] :hankey: Bugfix 
- [x] :pizza: New feature 
- [ ] :hammer: Refactoring 
- [ ] :book: CHANGELOG entry made
- [ ] :muscle: The behavior in this PR was *manually tested*
- [ ] :computer: The behavior change can be verified via automatic tests

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions)*

#### Related Information

*(Please reference tickets / specification etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added **Regression Tests**
  - [x] Added new C++ **Unit Tests**
  - [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

*(Include link to Jenkins run etc)*

### Documentation

> All new Features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the changelog so that
> developers and users have a concise overview. 

- [x] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)
- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/12421/

File: arangod/Aql/RemoteExecutor.cpp [hunk 1]
```diff
@@ -503,8 +503,8 @@ Result ExecutionBlockImpl<RemoteExecutor>::sendAsyncRequest(fuerte::RestVerb typ
   req->timeout(kDefaultTimeOutSecs);
   if (!_distributeId.empty()) {
     req->header.addMeta("x-shard-id", _distributeId);
```
Comment (hkernbach) [inline arangod/Aql/RemoteExecutor.cpp:505]: Maybe out of scope here but we should stick to static strings everywhere. As it is newly implemented in L507

File: arangod/Aql/RestAqlHandler.cpp [hunk 2]
```diff
@@ -715,7 +715,6 @@ RestStatus RestAqlHandler::handleUseQuery(std::string const& operation,
         return RestStatus::WAITING;
       }
     }
-    // Used in 3.4.0 onwards.
     answerBuilder.add("done", VPackValue(state == ExecutionState::DONE));
```
Comment (hkernbach) [inline arangod/Aql/RestAqlHandler.cpp:718]: same here

File: arangod/Aql/grammar.hpp [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/* A Bison parser, made by GNU Bison 3.7.2.  */
+/* A Bison parser, made by GNU Bison 3.7.1.  */
```
Comment (hkernbach) [inline arangod/Aql/grammar.hpp:1]: will this cause any problems?(different version used as file which is already checked in)

File: arangod/Cluster/ClusterMethods.cpp [hunk 4]
```diff
@@ -1889,16 +1891,23 @@ Result fetchEdgesFromEngines(transaction::Methods& trx,
...
       VPackSlice id = e.get(StaticStrings::IdString);
       if (!id.isString()) {
         // invalid id type
         LOG_TOPIC("a23b5", ERR, Logger::GRAPHS)
             << "got invalid edge id type: " << id.typeName();
         continue;
+      
+      }
+      if (result.capacity() == 0) {
+        result.reserve(16);
```
Comment (hkernbach) [inline arangod/Cluster/ClusterMethods.cpp:1907]: Just a question: Why 16 here?

File: arangod/Cluster/ClusterTraverser.cpp [hunk 5]
```diff
@@ -72,15 +73,15 @@ void ClusterTraverser::setStartVertex(std::string const& vid) {
...
+  if (!vertexMatchesConditions(s.stringRef(), 0)) {
     // Start vertex invalid
     _done = true;
     return;
   }
-
-  arangodb::velocypack::StringRef persId = traverserCache()->persistString(s);
-  _vertexGetter->reset(persId);
-  _enumerator->setStartVertex(persId);
+
```
Comment (hkernbach) [inline arangod/Cluster/ClusterTraverser.cpp:81]:
```suggestion
 
```

File: arangod/Aql/GraphNode.h [hunk 6]
```diff
@@ -58,6 +58,21 @@ namespace aql {
 class ExecutionEngine;
 
 class GraphNode : public ExecutionNode {
+ public:
+  struct InputVertex {
+    enum class Type { CONSTANT, REGISTER };
+    Type type;
+    // TODO make the following two a union instead
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/GraphNode.h:65]: This will be part of the Refactor, bot relevant here.

File: arangod/Aql/KShortestPathsNode.cpp [hunk 7]
```diff
@@ -166,11 +172,17 @@ KShortestPathsNode::~KShortestPathsNode() = default;
...
                                        arangodb::velocypack::Slice const& base)
     : GraphNode(plan, base),
+      _shortestPathType(arangodb::graph::ShortestPathType::Type::KShortestPaths),
       _pathOutVariable(nullptr),
       _inStartVariable(nullptr),
       _inTargetVariable(nullptr),
       _fromCondition(nullptr),
       _toCondition(nullptr) {
+  
+  if (base.hasKey("shortestPathType")) {
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/KShortestPathsNode.cpp:182]: Agree, will fix!

File: arangod/Aql/grammar.hpp (hunk 3 above)
Comment (mchacki) [reply arangod/Aql/grammar.hpp:1]: Good Catch!
AFAIK we rebuilt the generated files once before every official release to avoid potential issues.
We just check them in to be able to build everything if Bison/Flex are not installed, or outdated on the developers machine / You build without MaintainerMode. (Maybe also for Windows, but I am not sure about this)

So no I think in practice this does not matter.

File: arangod/Cluster/ClusterMethods.cpp (hunk 4 above)
Comment (mchacki) [reply arangod/Cluster/ClusterMethods.cpp:1921]: Let me rephrase:
Why at all?

there is a reserve on the exact number of results a couple of lines above, so this of no use.
I will remove

Comment (mchacki) [reply arangod/Cluster/ClusterMethods.cpp:1921]: (I think the initial intention was to avoid to small rescaling steps and 16 was an arbitrary number)

File: arangod/Aql/RestAqlHandler.cpp (hunk 2 above)
Comment (mchacki) [reply arangod/Aql/RestAqlHandler.cpp:715]: I agree this should be changed, but this is not related to this PR, so I would like to not include it here.

File: arangod/Graph/KPathFinder.cpp [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+  _cursor->readAll([&](EdgeDocumentToken&& eid, VPackSlice edge, size_t /*cursorIdx*/) -> void {
+    VertexRef id = _cache->persistString(([&]() -> auto {
+      if (edge.isString()) {
+        return VertexRef(edge);
+      } else {
+        VertexRef other(transaction::helpers::extractFromFromDocument(edge));
+        if (other == vertex) {
+          other = VertexRef(transaction::helpers::extractToFromDocument(edge));
+        }
+        return other;
```
Comment (jsteemann) [inline arangod/Graph/KPathFinder.cpp:204]: Still wondering if the following code is equivalent to the existing code...
```
diff --git a/arangod/Graph/KPathFinder.cpp b/arangod/Graph/KPathFinder.cpp
index 5ca733ab76..a893359cf0 100644
--- a/arangod/Graph/KPathFinder.cpp
+++ b/arangod/Graph/KPathFinder.cpp
@@ -188,11 +188,17 @@ auto KPathFinder::Ball::computeNeighbourhoodOfNextVertex(Ball const& other,
       if (edge.isString()) {
         return VertexRef(edge);
       } else {
-        VertexRef other(transaction::helpers::extractFromFromDocument(edge));
+        if (other._direction == Direction::FORWARD) {
+          return VertexRef(transaction::helpers::extractFromFromDocument(edge));
+        } else {
+          return VertexRef(transaction::helpers::extractToFromDocument(edge));
+        }
+        /*
         if (other == vertex) {
           other = VertexRef(transaction::helpers::extractToFromDocument(edge));
         }
         return other;
+        */
       }
     })()); 
```

File: arangod/RestHandler/RestBaseHandler.cpp [hunk 9]
```diff
@@ -49,9 +47,20 @@ RestBaseHandler::RestBaseHandler(application_features::ApplicationServer& server
...
+      // with authentication off or by super user. that means the caller either
+      // has full privileges anyway, or is the superuser, which is trusted.
+      strictValidation = false;
+    }
+  }
+
   try {
     success = true;
-    return _request->payload(true);
+    return _request->payload(strictValidation);
```
Comment (jsteemann) [inline arangod/RestHandler/RestBaseHandler.cpp:63]: This should have been removed from the PR, as discussed.

**Review by jsteemann — CHANGES_REQUESTED** [review]

File: arangod/Graph/KPathFinder.cpp (hunk 8 above)
Comment (mchacki) [reply arangod/Graph/KPathFinder.cpp:204]: I think it is not.
The "FORWARD" here is not defining that we go from _from to _to.
It is defining that we go in the direction the user has written.
e.g. OUTBOUND (which is _from to _to) but if the user has written INBOUND the above suggestion would return the wrong vertex.

We plan on improving this in a different PR once and for all.

Comment (jsteemann) [reply arangod/Graph/KPathFinder.cpp:204]: Vale

File: arangod/RestHandler/RestBaseHandler.cpp (hunk 9 above)
Comment (mchacki) [reply arangod/RestHandler/RestBaseHandler.cpp]: Correct, thanks for pointing this out agin.
Is now removed.

File: CHANGELOG [hunk 10]
```diff
@@ -1,6 +1,21 @@
 devel
 -----
 
+* Added new graph method K_PATHS to AQL. This will enumerate all paths between a
+  source and a target vertex that match the given length.
+  e.g.
```
Comment (jsteemann) [inline CHANGELOG:6]:
```suggestion
  For example, the query
```

File: CHANGELOG [hunk 11]
```diff
@@ -1,6 +1,21 @@
...
+  ```
+    FOR path IN 2..4 OUTBOUND K_PATHS "v/source" TO "v/target" GRAPH "g"
+      RETURN path
+  ```
+  Will yield all paths in format
```
Comment (jsteemann) [inline CHANGELOG:11]:
```suggestion
  will yield all paths in format
```

File: CHANGELOG [hunk 12]
```diff
@@ -1,6 +1,21 @@
...
+    {
+      vertices: [v/source, ... , v/target],
+      edges: [v/source -> v/1, ...,  v/n -> v/target
+    }
+  that have length exactly 2 or 3 or 4, start at v/source and end at v/target.
+  The ordering of those paths is not guaranteed.
```
Comment (jsteemann) [inline CHANGELOG:17]:
```suggestion
  The order of those paths in the result set is not guaranteed.
```

File: arangod/Cluster/ClusterMethods.h [hunk 13]
```diff
@@ -51,16 +51,22 @@ class ClusterTraverserCache;
...
-struct ClusterCommResult;
 class ClusterFeature;
 struct OperationOptions;
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief creates a copy of all HTTP headers to forward
+////////////////////////////////////////////////////////////////////////////////
+
+network::Headers getForwardableRequestHeaders(GeneralRequest*);
+
```
Comment (jsteemann) [inline arangod/Cluster/ClusterMethods.h:69]: I don't think we should add this back. This is from 3.6 and should be unneeded now.
```suggestion
```

File: arangod/Graph/ClusterTraverserCache.cpp [hunk 14]
```diff
@@ -61,7 +62,9 @@ aql::AqlValue ClusterTraverserCache::fetchVertexAqlResult(arangodb::velocypack::
   // FIXME: this is only used for ShortestPath, where the shortestpath stuff
   // uses _edges to store its vertices
 
-  auto it = _cache.find(id);
+  // There will be no idString of length above uint32_t
+  auto it = _cache.find(
+      arangodb::velocypack::HashedStringRef(id.data(), static_cast<uint32_t>(id.length())));
 
   if (it != _cache.end()) {
     // FIXME: the ClusterTraverserCache lifetime is shorter then the query
```
Comment (jsteemann) [inline arangod/Graph/ClusterTraverserCache.cpp:70]:
```suggestion
    // FIXME: the ClusterTraverserCache lifetime is shorter than the query
```

**Review by jsteemann — APPROVED** [review]

**Review by hkernbach — APPROVED** [review]

**Comment by mchacki** [comment]
Jenkins Blue:
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/12467/
